src/lazy/binary/raw/v1_1/e_expression.rs

```diff
@@ -428,28 +468,48 @@ impl<'top> HasSpan<'top> for BinaryEExpArgGroup<'top> {
 #[derive(Debug, Copy, Clone)]
 pub struct BinaryEExpArgGroupIterator<'top> {
     parameter: &'top Parameter,
-    remaining_args_buffer: BinaryBuffer<'top>,
+    source: BinaryEExpArgGroupIteratorSource<'top>,
+}
+
+#[derive(Debug, Copy, Clone)]
+enum BinaryEExpArgGroupIteratorSource<'top> {
+    /// The iterator is parsing expressions from a slice of the input buffer
+    Input(BinaryBuffer<'top>),
+    /// The iterator is iterating over the bump-allocated cache of previously parsed expressions.
+    Cache(&'top [LazyRawValueExpr<'top, BinaryEncoding_1_1>]),
```

Comment:
So, the name and doc comments suggest to me that these expressions could be shared somehow. (Maybe there's a variable that's used more than once...?) Yet, further down, I see that this is being mutated. Is this actually a slice that points to a cache that is owned somewhere else? And it's effectively just the start and length of the slice that is updated?

Are any of those cached expressions stateful? Are we sure that there's no possibility of interference between different use-sites of the same cached expressions?

Reply:
Both of `BinaryEExpArgGroupIteratorSource`'s variants refer to data stored elsewhere with a lifetime of `'top`:
* In the `Input` case, the `BinaryBuffer` holds a slice of the input buffer, which is frozen for the duration of `'top`.
* In the `Cache` case, the reader bump-allocated a `&'top [...]` in which to store raw lexed expressions it found as it traversed a delimited group. Data stored in the bump is guaranteed to outlive `'top`.

`BinaryEExpArgGroupIteratorSource` instances are immutable, but they are each wrapped by a `BinaryEExpArgGroupIterator` that _is_ mutable. Each time the iterator's `next()` method is called, it either advances the slice of the source's `Input` variant that it holds or it pops the head off of the slice of the `Cache` variant that it holds.

Both the iterator and its source are trivially copy-able, so you could easily/safely make several iterators traversing the expression group at once. Each iterator would mutate its own state as it traversed the immutable source data.

Is that helpful?